Support for continuing to scan the same string.

kbloom opened the pull request.

kbloom committed:
Add a module file, so that I can run "go test" with modern versions of Go.

ghemawat commented on examples_test.go:
```diff
@@ -67,3 +67,25 @@ func ExampleScan_parseDuration() {
+		)
+		err := re.Scan(r, line, &position, &host, &port)
+		if err != nil {
+			break
```
Do we need to document what error is returned on end of input as opposed to a sub-object parse failure? (It might be a bug if a group matches something that cannot be converted to the corresponding result, but it would be expected in such a repeated loop for us to not find a match at all.)

Once we have such a distinction, we can break the loop on the no match case, and panic on other errors.

I am unsure of what the special "no match" error should be. Maybe just re.NoMatch? So the code here will become:

```
if err := re.Scan(...); err != nil {
  if errors.Is(err, re.NoMatch) {
    break
  }
  panic(err)
}
```

ghemawat commented on go.mod:
```diff
@@ -0,0 +1,3 @@
+module github.com/ghemawat/re
+
+go 1.19
```
Do we need to require 1.19? Maybe 1.18 would be better?

kbloom replied:
Set to Go 1.11, because that's the first version that introduced go.mod (and your code is way older than that release)

ghemawat commented on re.go:
```diff
@@ -29,6 +29,14 @@ import (
 	"strconv"
 )
 
+// A pointer to a Position can be passed as the third argument to Scan to
```
Go style suggests the comment starts with the name of the thing being documented. So for example, this could be changed to something like:

> Position is a special type designed to be passed via pointer to re.Scan. re.Scan will store the starting and ending offsets of the corresponding regular expression capture group into the Position.
>
> This type is most helpful as a way to capture where to start the next scan when scanning repeatedly through some data.

ghemawat commented on re.go:
```diff
@@ -29,6 +29,14 @@ import (
 
+// A pointer to a Position can be passed as the third argument to Scan to
+// record position information for the match.  This can be used in a loop to
+// continue scanning from the end of previous first match.
```
Note that the "third argument" above may not be right since Scan is variadic. People might place the *Position at the end of the argument list. E.g.,

exp := regexp.MustCompile(`(\S+):(\d+)()`)
re.Scan(data, &host, &port, &pos)

kbloom replied:
I'm essentially proposing an API where there:

1. 2 required arguments: needle (the regexp) and haystack
2. An optional argument (a Range describing the whole match)
3. Variable arguments that come after the optional argument.

Of course, Go doesn't suppport optional arguments, but we can fake it since our variable arguments are interface{}

ghemawat replied:
I see your point that many users will directly want the full range of the match. However I am still favoring the alternate API: re.Range behaves like any other destination argument to re.Scan: it is filled with the sub-portion of the match corresponding to a capture group. A few reasons for this:

1. We don't introduce the complexity of a special parameter that can only occur in a certain position and disturbs the expected position of other parameters. We stick with the simple rule that the ith parameter is fed the ith capture group.
2. Treating re.Range the same as other parameters allows some extra generality: the user can get intermediate offsets from within the middle of the match if they want.

The downside of my proposal is that the user has to change the regexp, but the regexp has to be closely tied to the re.Scan call anyway, so I am not that concerned about this requirement and on balance would like to keep re.Range like other captured parameters.

kbloom replied:
OK. Done.

ghemawat commented on examples_test.go:
```diff
@@ -67,3 +67,25 @@ func ExampleScan_parseDuration() {
+			host     string
+			port     int
+		)
+		err := re.Scan(r, line, &position, &host, &port)
```
My intention with this suggestion was that re.Position would not be special. It can occur anywhere in the list of args and matches a capture group. So for example, a user could place an empty capture group at the end:

```
   `(\S+):(\d+)()` // args will be &host, &port, &pos
```

Or capture the whole thing:

```
   `((\S+):(\d+))  // args will be &pos, &host, &port
```

kbloom replied:
My intention was that you could capture Range of the whole regexp without having to actually change how you wrote the regexp.

It might also be useful to capture Range of individual matching groups, but we can do that by supporting additional types of parsing functions. If we add support for func([]byte, Range) error, then we can define our first custom parsing function:

```
func GroupPosition(value interface{}, range *Range) func([]byte, Range) error {
  return func(content []byte, r Range) error {
   *range = r
   return assign(value, content)
  }
}

```
and get the range for a matching group as follows:

```
exp := regexp.MustCompile(`(\S+):(\d+)`)
re.Scan(exp, data, re.GroupPosition(&host, &hostRange), &port)
```

(I haven't thought much about the name for GroupPosition yet)

kbloom replied:
I've implemented `func(content []byte, r Range) error`, but haven't written the GroupPosition function yet. I'm not sure this is a common thing to need, so I'm not sure whether we should provide the GroupPosition function by default. Maybe we should use this support to implement a `Maybe`, though, which would have similar functionality to C++ RE2's support of std::optional.

ghemawat replied:
Please see my other comment. I would like add just one new type that captures a sub-range of the match like other parameters.

kbloom replied:
So in the version I'm about to push

- Passing *re.Span captures the position information for the corresponding subgroup.
- The third argument is no longer special.
- Despite the fact that the third argument is no longer special, the doucmentation for re.Span says that for scanning across an input multiple times, you should wrap the entire regexp in parentheses, and pass *re.Span as the third argument. There are other ways to do this, like having an empty parentheses at the end, but IMO, it's good to recommend one in particular.
- func(content []byte, span Span) error is still a thing. I'll work on a followup that adds some formatting functions that make use of it.

ghemawat commented on examples_test.go:
```diff
@@ -67,3 +67,25 @@ func ExampleScan_parseDuration() {
+
+	for {
+		var (
+			position re.Position
```
Looking at this in context, perhaps re.Range would be a better name (Position implies a single int)

kbloom replied:
I've adopted this change, but lowercase `range` is a reserved word in Go, so `var range Range` isn't going to fly. Throughtout my code, I've been using `rng` to deal with this, but if that's not ideal, we might reconsider naming again.

ghemawat replied:
Good point. WDYT about re.Extent or re.Span?

kbloom replied:
I chose re.Span.

ghemawat commented on re_test.go:
```diff
@@ -202,6 +202,52 @@ func TestReFunc(t *testing.T) {
+	var port int
+	if err := re.Scan(regexp.MustCompile(hp), bytes, &pos, &host, &port); err != nil {
+		t.Fatalf("First match: unexpected error: %s", err)
+	} else {
```
Drop the else and move the rest of the body out (Go style suggests not indenting after something like t.Fatal that has already returned from this function).

ghemawat commented on re_test.go:
```diff
@@ -202,6 +202,52 @@ func TestReFunc(t *testing.T) {
 	}
 }
 
+func TestRePosition(t *testing.T) {
```
Not: rename to TestPosition

ghemawat commented on re_test.go:
```diff
@@ -202,6 +202,52 @@ func TestReFunc(t *testing.T) {
+		}
+	}
+
+}
```
Table driven test may  be less repetitive and easier to extend:

```
type testMatch struct {
   Pos re.Position
   Host string
   Port int
}
type testCase struct {
   input string
   matches []testMatch
}

for _, c := range []testCase{
   {"...", testMatch{...}},
   ...
} {
   t.Run(c.input, func(t *testing.T) {
     loop that collects a slice of matches and then uses reflect.DeepEqual to compare
   })
}

ghemawat reviewed (changes requested):
Thanks!

I am being picky, but mostly because  the interesting part here is the API and it will be hard to fix it later if we get things wrong.

kbloom commented:
PTAL.

ghemawat commented on examples_test.go:
```diff
@@ -67,3 +67,31 @@ func ExampleScan_parseDuration() {
+			port int
+		)
+		err := re.Scan(r, line, &rng, &host, &port)
+		if _, notFound := err.(re.NotFoundError); notFound {
```
I think you can change this to:
   if errors.Is(re.NotFoundError) {  ... }

I think this API showed up in Go 1.13.

ghemawat commented on go.mod:
```diff
@@ -0,0 +1,3 @@
+module github.com/ghemawat/re
+
+go 1.11
```
Might need 1.13 for errors.Is

kbloom replied:
Done.

ghemawat commented on re.go:
```diff
@@ -29,6 +29,41 @@ import (
+}
+
+// ParseError is the error retured by re.Scan when a submatch couldn't be
+// parsed in the format expected by the corresponding submatch output.
```
I think we should leave ParseError out. When a user provides a parsing function they will return their own error and there will be a resulting confusion as to whether or not the error is theirs or a ParseError.

I think just adding NotFoundError is sufficient for now.

kbloom replied:
Done.

ghemawat commented on re.go:
```diff
@@ -29,6 +29,41 @@ import (
+// the regexp didn't match the string.
+type NotFoundError struct {
+	re    *regexp.Regexp
+	input []byte
```
input may be huge and errors can be saved away and kept around for a while. We should drop input from the error.

kbloom replied:
Done.

ghemawat commented on re.go:
```diff
@@ -29,6 +29,41 @@ import (
+}
+
+func (e NotFoundError) Error() string {
+	return fmt.Sprintf(`re.Scan: could not find "%s" in "%s"`,
```
fmt's documentation indicates that we should %v, not %s for printing the regexp.

ghemawat commented on re.go:
```diff
@@ -29,6 +29,41 @@ import (
+
+// NotFoundError is the error returned by re.Scan when
+// the regexp didn't match the string.
+type NotFoundError struct {
```
If you change the exported thing to be a var, callers can do:
   if errors.Is(re.NotFound) { ... }

Also, maybe we can drop the Error suffix so this could be:

var NotFound = errors.New("not found")

And then when returning an error we can just do:
  return fmt.Errorf("regular expression %q: %w", re, NotFound)
(intentionally using %q to get quoting of spaces etc. in the regexp)

Note that this is also the approach often used in the stdlib (e.g., io).

kbloom replied:
Done.

ghemawat commented on re.go:
```diff
@@ -111,7 +167,7 @@ func assign(r interface{}, b []byte) error {
 		}
 		if int64(int(i)) != i {
-			return parseError("out of range for int", b)
+			return fmt.Errorf("out of range for int: %s", b)
```
I would suggest reverting back to the old parseError approach (see comments about ParseError earlier)

kbloom replied:
Done.

ghemawat commented on re_test.go:
```diff
@@ -202,6 +202,68 @@ func TestReFunc(t *testing.T) {
+					Port:  2345,
+				},
+			},
+		},
```
Can you add a test case where there is no match.

kbloom replied:
Done

ghemawat commented on re_test.go:
```diff
@@ -202,6 +202,68 @@ func TestReFunc(t *testing.T) {
+					t.Errorf("Scan attempt %d: port = %d, want %d", idx, port, match.Port)
+				}
+				input = input[rng.End:]
+			}
```
How about another Scan call and a check that it returns a not-found error.

kbloom replied:
Done.

ghemawat commented on re.go:
```diff
@@ -29,6 +29,41 @@ import (
+// passed as the third parameter to Scan, Scan will record position information
+// for the match.  This can be used in a loop to continue scanning from the end
+// of previous first match.
+type Range struct {
```
I like your suggestion about picking a different name so users don't have to avoid the obvious variable name.

Maybe re.Extent or re.Span?

kbloom replied:
I chose Span.

kbloom committed:
add re.Span, which returns the position of the full match.

This allows the caller to iteratively scan across the input until the string is
exhausted.

kbloom committed:
Introduce NotFound error.

Callers can look for NotFound using errors.Is to figure out when Scan could not
find the regexp in the input.  Callers using Scan to scan for multiple matches
across a string can single out this error to identify when to stop scanning.

Because we do this by wrapping the NotFound error (and expect the caller to use
errors.Is to match it), this now requires Go 1.13. go.mod has been updated
accordingly.

kbloom commented:
OK. I've addressed your comments

ghemawat commented on re.go:
```diff
@@ -23,12 +23,31 @@ custom parsing.
 	"strconv"
 )
 
+// Span is a special type designed to be passed via pointer to Scan.  re.Scan
```
Nice comment.

I agree with the recommendation for suggesting to capture the entire regexp, but perhaps reword to something like:

This type can be placed anywhere within the list of arguments to scan, but the most typical usage is to find the entire extent of the match, which can be achieved by placing it third (immediately after the regular expression and input), and wrapping the entire regexp in parentheses so that the Span is filled with the extent of the entire match.

kbloom replied:
Done.

ghemawat commented on re.go:
```diff
@@ -64,6 +83,12 @@ import (
 // (normally Scan would treat such as a number as octal); or parsing
 // an otherwise unsupported type like time.Duration.
 //
+// func([]byte, Span) error: This is another supported form of custom
```
Drop this for now.

Passing &Span suffices for uncommon uses (people can run their custom code later on Span if they want).

kbloom replied:
Done.

ghemawat commented on re_test.go:
```diff
@@ -202,6 +203,79 @@ func TestReFunc(t *testing.T) {
+
+			err := re.Scan(pattern, input, &span, &host, &port)
+			if !errors.Is(err, re.NotFound) {
+				t.Errorf("Scan attempt %d: Error was %v, want an error that wraps %v", len(c.matches), err, re.NotFound)
```
Maybe change the last %v to %T so we print the type that was expected, not an instance of that type.

kbloom replied:
Because we created the error with errors.New, the type really isn't interesting. The text is the only interesting part.

kbloom commented:
Done.
